noir_stdlib/src/meta/map.nr

```diff
@@ -0,0 +1,458 @@
+
+impl<K, V> Slot<K, V> {
+    fn is_valid(self) -> bool {
+        !self._is_deleted & self._key_value.is_some()
+    }
+
+    fn is_available(self) -> bool {
+        self._is_deleted | self._key_value.is_none()
+    }
+
+    fn key_value(self) -> Option<(K, V)> {
+        self._key_value
+    }
+
+    fn key_value_unchecked(self) -> (K, V) {
+        self._key_value.unwrap_unchecked()
+    }
+
+    fn set(&mut self, key: K, value: V) {
+        self._key_value = Option::some((key, value));
+        self._is_deleted = false;
+    }
+
+    // Shall not override `_key_value` with Option::none(),
+    // because we must be able to differentiate empty 
+    // and deleted slots for lookup.
+    fn mark_deleted(&mut self) {
+        self._is_deleted = true;
+    }
+}
+
+// While conducting lookup, we iterate attempt from 0 to N - 1 due to heuristic,
+// that if we have went that far without finding desired, 
+// it is very unlikely to be after - performance will be heavily degraded.
+impl<K, V, B, H> CtHashMap<K, V, B> {
+    // Creates a new instance of CtHashMap with specified BuildHasher.
+    // docs:start:with_hasher
+    pub fn with_hasher(_build_hasher: B) -> Self
+    where
+        B: BuildHasher<H> {
+        // docs:end:with_hasher
+        let _table = &[Slot::default()];
+        let _len = 0;
+        Self { _table, _len, _build_hasher }
+    }
+
+    pub fn with_hasher_and_capacity(_build_hasher: B, capacity: u32) -> Self
+    where
+        B: BuildHasher<H> {
+        // docs:end:with_hasher
+        let mut _table = &[];
+        for _ in 0..capacity {
+            _table = _table.push_back(Slot::default());
+        }
+        let _len = 0;
+        Self { _table, _len, _build_hasher }
+    }
+
+    // Clears the map, removing all key-value entries.
+    // docs:start:clear
+    pub fn clear(&mut self) {
+        // docs:end:clear
+        self._table = &[Slot::default()];
+        self._len = 0;
+    }
+
+    // Returns true if the map contains a value for the specified key.
+    // docs:start:contains_key
+    pub fn contains_key(
+        self,
+        key: K
+    ) -> bool
+    where
+        K: Hash + Eq,
+        B: BuildHasher<H>,
+        H: Hasher {
+        // docs:end:contains_key
+        self.get(key).is_some()
+    }
+
+    // Returns true if the map contains no elements.
+    // docs:start:is_empty
+    pub fn is_empty(self) -> bool {
+        // docs:end:is_empty
+        self._len == 0
+    }
+
+    // Returns a BoundedVec of all valid entries in this CtHashMap.
+    // The length of the returned vector will always match the length of this CtHashMap.
+    // docs:start:entries
+    pub fn entries(self) -> [(K, V)] {
+        // docs:end:entries
+        let mut entries = &[];
+
+        for slot in self._table {
+            if slot.is_valid() {
+                // SAFETY: slot.is_valid() should ensure there is a valid key-value pairing here
+                let key_value = slot.key_value().unwrap_unchecked();
```

Comment:
Why not use `key_value_unchecked()` here?

Reply:
Not sure - I copied most of this code from the existing HashMap